core/src/main/scala/com/criteo/cuttle/Workflow.scala

```diff
@@ -16,30 +17,144 @@ trait Workflow[S <: Scheduling] {
+
+    def dfs(node: NodeWithMetadata): Unit = {
+      stack.push(node)
+      nodesInStack += node
+      time += 1
+      node.discoveryTime = time
+      node.earliestAncestor = time
+      node.children.foreach { child =>
+        if (!child.visited) {
+          // (node -> child) is a tree edge
+          dfs(child)
+          node.earliestAncestor = Math.min(node.earliestAncestor, child.earliestAncestor)
+        }
+        else if (nodesInStack.contains(child)) {
+            // (node -> child) is a back edge
+            node.earliestAncestor = Math.min(node.earliestAncestor, child.discoveryTime)
+        }
+        // Otherwise 'child' was already added to a SCC
+      }
+
+      // Depth-first search tree of 'node' full explored
+      if (node.discoveryTime == node.earliestAncestor) {
+        // 'node' is the head node of a SCC. The nodes in the 'stack' above it are part of the same SCC
+        val scc = mutable.ListBuffer.empty[Job[S]]
+        while (stack.top != node) {
+          scc += stack.top.node
+          nodesInStack -= stack.top
+          stack.pop()
+        }
+        scc += stack.top.node
+        nodesInStack -= stack.top
+        stack.pop()
+        stronglyConnectedComponents += scc.toList
+      }
+    }
+
+    val nodesWithMetadata = vertices.map(job => NodeWithMetadata(job))
+      .map(node => node.node -> node).toMap
+    edges.foreach { case(child, parent, _) => nodesWithMetadata(parent).addChild(nodesWithMetadata(child)) }
+
+    nodesWithMetadata.values.foreach { node =>
+      if (!node.visited) dfs(node)
+    }
+
+    stronglyConnectedComponents.filter(_.size >= 2).toList
+  }
+
+  /**
+    * @param edgeVisitor method called for each edge (parent node, child node) visited by Kahn's algorithm
+    * @return the jobs in the workflow sorted in topological order, None if the graph contains a cycle
+    * */
+  def topologicalSort(edgeVisitor: (Job[S], Job[S]) => Unit = (_, _) => ()): Option[List[Job[S]]] = {
+    val topologicalOrder = mutable.ListBuffer.empty[Job[S]]
+
+    case class NodeWithMetadata(node: Job[S], children: mutable.Set[NodeWithMetadata] = mutable.Set.empty) {
+      private var numIncomingEdges: Long = 0
+
+      def addChild(child: NodeWithMetadata): Boolean = {
+        child.numIncomingEdges += 1
+        children.add(child)
       }
+      def removeChild(child: NodeWithMetadata): Boolean = {
+        child.numIncomingEdges -= 1
+        children.remove(child)
+      }
+      /** @return true if the node has no incoming edges */
+      def isOrphanNode: Boolean = numIncomingEdges == 0
+
+      // Override hash computation to prevent stack overflow when trying to compute the hash of a node using the
+      // default hash method on workflows with cycles
+      override def hashCode(): Int = node.hashCode()
     }
 
-    require(edges.isEmpty, "Workflow has at least one cycle")
-    result.toList
+    val nodesWithMetadata = vertices.map(job => NodeWithMetadata(job))
```

Comment:
I think `val nodesWithMetadata = vertices.map(job => (job, NodeWithMetadata(job))).toMap` is more readable...

Reply:
Sure, will do